invenio_rdm_records/services/access/service.py

```diff
@@ -0,0 +1,761 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020-2021 CERN.
+# Copyright (C) 2020-2021 Northwestern University.
+# Copyright (C) 2021 TU Wien.
+# Copyright (C) 2023 Graz University of Technology.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""RDM record access settings service."""
+
+from datetime import datetime, timedelta
+
+import arrow
+from flask import current_app, url_for
+from invenio_access.permissions import authenticated_user, system_identity
+from invenio_drafts_resources.services.records import RecordService
+from invenio_i18n import lazy_gettext as _
+from invenio_records_resources.services.errors import PermissionDeniedError
+from invenio_records_resources.services.records.schema import ServiceSchemaWrapper
+from invenio_records_resources.services.uow import RecordCommitOp, unit_of_work
+from invenio_requests.proxies import current_requests_service
+from invenio_users_resources.proxies import current_user_resources
+from marshmallow.exceptions import ValidationError
+from sqlalchemy.orm.exc import NoResultFound
+
+from ...requests.access import AccessRequestToken, GuestAccessRequest, UserAccessRequest
+from ...requests.access.requests import EmailOp
+from ...secret_links.errors import InvalidPermissionLevelError
+from ..errors import DuplicateAccessRequestError
+from ..results import GrantSubjectExpandableField
+
+
+class RecordAccessService(RecordService):
+    """RDM Secret Link service."""
+
+    def link_result_item(self, *args, **kwargs):
+        """Create a new instance of the resource unit."""
+        return self.config.link_result_item_cls(*args, **kwargs)
+
+    def link_result_list(self, *args, **kwargs):
+        """Create a new instance of the resource list."""
+        return self.config.link_result_list_cls(*args, **kwargs)
+
+    def grant_result_item(self, *args, **kwargs):
+        """Create a new instance of the resource unit."""
+        # TODO it seems like user profiles are expanded even if they are hidden
```

Comment:
As soon as you have a `grant` for a certain `user` in the record (i.e. "you shared access to the record with that user's account"), it seems like you will be able to see that user's profile information by **expanding** the grant (`?expand=1`), no matter if that user has set their profile visibility to public or restricted.
This may leak some information about the user to you (as the record owner) that the user wasn't comfortable with sharing.

Note that the current workflow doesn't allow you to share access to a record with a user unless you can read their profile though.

So this may circumvent a user's privacy settings if they decided to hide their profile only after you've shared access to a record with them!